### ISSUE TITLE
Multiple ESLint failures

### ISSUE THREAD

**Issue by MikeMcC399** [cypress-io/cypress-realworld-app#1547]
## Issue

If the repo is manually linted with ESLint then multiple issues are reported. Violations of the following rules are reported:

### array-callback-return

Rule: [array-callback-return](https://eslint.org/docs/latest/rules/array-callback-return)

Violations in:

- `cypress-realworld-app/backend/database.ts`

### cypress/no-unnecessary-waiting

Rule: [cypress/no-unnecessary-waiting](https://github.com/cypress-io/eslint-plugin-cypress/blob/master/docs/rules/no-unnecessary-waiting.md)

Violations in:

- `cypress-realworld-app/cypress/tests/ui/transaction-feeds.spec.ts`

### cypress/unsafe-to-chain-command

Rule: [cypress/unsafe-to-chain-command](https://github.com/cypress-io/eslint-plugin-cypress/blob/master/docs/rules/unsafe-to-chain-command.md)

Violations in:

- `cypress-realworld-app/cypress/tests/ui/auth.spec.ts`
- `cypress-realworld-app/cypress/tests/ui/bankaccounts.spec.ts`
- `cypress-realworld-app/cypress/tests/ui/new-transaction.spec.ts`
- `cypress-realworld-app/cypress/tests/ui/transaction-feeds.spec.ts`
- `cypress-realworld-app/cypress/tests/ui/user-settings.spec.ts`

### @typescript-eslint/no-unused-expressions

Rule: [@typescript-eslint/no-unused-expressions](https://typescript-eslint.io/rules/no-unused-expressions)

Violations in:

`cypress-realworld-app/cypress/tests/ui/transaction-feeds.spec.ts`

### @typescript-eslint/no-unused-vars

Rule: [@typescript-eslint/no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars)

Violations in:

- `cypress-realworld-app/cypress/global.d.ts`

## Steps to reproduce

Ubuntu `22.04.5` LTS, Node.js `20.12.0` according to [.node-version](https://github.com/cypress-io/cypress-realworld-app/blob/develop/.node-version)

Execute

```shell
npm install yarn@latest -g
git clone https://github.com/cypress-io/cypress-realworld-app
cd cypress-realworld-app
n auto # only if using n as node.js version manager
yarn
yarn eslint .
```

## Logs

<details>
<summary>Logs for "yarn eslint ."</summary>
$ yarn eslint .
yarn run v1.22.19
$ /home/mike/github/tmp/cypress-realworld-app/node_modules/.bin/eslint .
Browserslist: caniuse-lite is outdated. Please run:
  npx update-browserslist-db@latest
  Why you should do it regularly: https://github.com/browserslist/update-db#readme

/home/mike/github/tmp/cypress-realworld-app/backend/database.ts
  768:57  warning  Array.prototype.flatMap() expects a value to be returned at the end of arrow function  array-callback-return

/home/mike/github/tmp/cypress-realworld-app/cypress/global.d.ts
  10:5  warning  'User' is defined but never used          @typescript-eslint/no-unused-vars
  11:5  warning  'BankAccount' is defined but never used   @typescript-eslint/no-unused-vars
  12:5  warning  'Like' is defined but never used          @typescript-eslint/no-unused-vars
  13:5  warning  'Comment' is defined but never used       @typescript-eslint/no-unused-vars
  14:5  warning  'Transaction' is defined but never used   @typescript-eslint/no-unused-vars
  15:5  warning  'BankTransfer' is defined but never used  @typescript-eslint/no-unused-vars
  16:5  warning  'Contact' is defined but never used       @typescript-eslint/no-unused-vars

/home/mike/github/tmp/cypress-realworld-app/cypress/tests/ui/auth.spec.ts
  114:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  114:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  118:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  133:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  133:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  136:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  136:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  139:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  139:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  142:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  142:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  145:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command

/home/mike/github/tmp/cypress-realworld-app/cypress/tests/ui/bankaccounts.spec.ts
   72:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
   72:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
   77:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
   84:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
   90:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
   96:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  101:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  107:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  113:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  118:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  122:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command

/home/mike/github/tmp/cypress-realworld-app/cypress/tests/ui/new-transaction.spec.ts
   79:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  114:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  126:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  126:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  131:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  131:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command

/home/mike/github/tmp/cypress-realworld-app/cypress/tests/ui/transaction-feeds.spec.ts
   76:9   error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  123:13  error  Expected an assignment or function call and instead saw an expression                                                                                                @typescript-eslint/no-unused-expressions
  177:9   error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  191:11  error  Do not wait for arbitrary time periods                                                                                                                               cypress/no-unnecessary-waiting
  236:11  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  308:9   error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  406:11  error  Expected an assignment or function call and instead saw an expression                                                                                                @typescript-eslint/no-unused-expressions
  427:11  error  Expected an assignment or function call and instead saw an expression                                                                                                @typescript-eslint/no-unused-expressions

/home/mike/github/tmp/cypress-realworld-app/cypress/tests/ui/user-settings.spec.ts
  32:7  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  32:7  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  38:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  38:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  43:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  48:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  48:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  53:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  63:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  64:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  65:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  66:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command
  66:5  error  It is unsafe to chain further commands that rely on the subject after this command. It is best to split the chain, chaining again from `cy.` in a next command line  cypress/unsafe-to-chain-command

✖ 58 problems (50 errors, 8 warnings)

One of your dependencies, babel-preset-react-app, is importing the
"@babel/plugin-proposal-private-property-in-object" package without
declaring it in its dependencies. This is currently working because
"@babel/plugin-proposal-private-property-in-object" is already in your
node_modules folder for unrelated reasons, but it may break at any time.

babel-preset-react-app is part of the create-react-app project, which
is not maintianed anymore. It is thus unlikely that this bug will
ever be fixed. Add "@babel/plugin-proposal-private-property-in-object" to
your devDependencies to work around this error. This will make this message
go away.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</details>

## Suggestions

Each of the failures needs to be examined to decide whether they are valid/fixable, valid/not-fixable or false positives.

ESLint provides different options for disabling certain rules globally, individually for a source file, a section of a source file or a single line of a source file. This may be a temporary or permanent solution for each failure.

### Script `lint`

The script `lint` is inactive as far as ESLint is concerned. Since no files are specified, it lints zero files, which is probably the reason that the linting failures have not so far been noticed.

This should be logged as a separate issue. It would however be better to address the lint failures first, before enabling the script to work.

### ESLint configuration

The ESLint configuration is contained in the `eslintConfig` property of [package.json](https://github.com/cypress-io/cypress-realworld-app/blob/develop/package.json). This ~~is~~ was allowed in [ESLint 8.x](https://eslint.org/docs/v8.x/use/configure/configuration-files) (version now unsupported) and deprecated in [ESLint 9.x](https://eslint.org/docs/latest/use/configure/configuration-files-deprecated).

It may be advantageous to update the configuration to use an `eslint.config.*js` format in parallel to resolving linting failures.

- Edit: since Oct 5, 2024, ESLint `v8`, used in this repo, is unsupported. See https://eslint.org/version-support/ and https://github.com/cypress-io/cypress-realworld-app/issues/1616

**Comment by MikeMcC399** [comment]
After merge of
- #1641 
- #1642 
- #1643

this issue should be closed and open aspects from the issue transferred to a new issue

**Comment by MikeMcC399** [comment]
## Status

Using Ubuntu `24.04.2` LTS with Node.js `22.15.0` according to [.node-version](https://github.com/cypress-io/cypress-realworld-app/blob/develop/.node-version)

and executing

```shell
npm install yarn@latest -g
git clone https://github.com/cypress-io/cypress-realworld-app
cd cypress-realworld-app
n auto # only if using n as node.js version manager
git clean -xfd
yarn
yarn eslint
```

results in

### Yarn installation log

```text
$ yarn
yarn install v1.22.22
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
[4/5] Linking dependencies...
warning " > react-infinite-calendar@2.3.1" has incorrect peer dependency "react@^15.3.0 || ^16.0.0-alpha".
warning "react-infinite-calendar > react-tiny-virtual-list@2.2.0" has incorrect peer dependency "react@15.x || 16.x".
warning "react-infinite-calendar > react-transition-group@1.2.1" has incorrect peer dependency "react@^15.0.0 || ^16.0.0".
warning "react-infinite-calendar > react-transition-group@1.2.1" has incorrect peer dependency "react-dom@^15.0.0 || ^16.0.0".
warning "react-infinite-calendar > recompose@0.22.0" has incorrect peer dependency "react@^0.14.0 || ^15.0.0".
warning " > @cypress/code-coverage@3.13.12" has unmet peer dependency "@babel/preset-env@^7.0.0".
warning " > @cypress/code-coverage@3.13.12" has unmet peer dependency "babel-loader@^8.3 || ^9".
warning "@cypress/code-coverage > @cypress/webpack-preprocessor@6.0.2" has unmet peer dependency "@babel/preset-env@^7.0.0".
warning "@cypress/code-coverage > @cypress/webpack-preprocessor@6.0.2" has unmet peer dependency "babel-loader@^8.3 || ^9".
warning Workspaces can only be enabled in private projects.
warning Workspaces can only be enabled in private projects.
[5/5] Building fresh packages...
$ husky install && patch-package
husky - Git hooks installed
patch-package 7.0.2
Applying patches...
react-virtualized@9.22.5 ✔
Done in 319.26s.
```

### ESLint log

```text
$ yarn eslint
yarn run v1.22.22
$ /home/mike/github/cypress-io/cypress-realworld-app/node_modules/.bin/eslint

/home/mike/github/cypress-io/cypress-realworld-app/backend/types.ts
  1:1  warning  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/no-unused-vars')

✖ 1 problem (0 errors, 1 warning)
  0 errors and 1 warning potentially fixable with the `--fix` option.

Done in 3.96s.
```

**Comment by MikeMcC399** [comment]
Given that major linting issues have been remedied and linting now uses supported versions, this issue can now be closed. Actions remain to review disabled rules. This is not explicitly tracked in any GitHub issue at this time.

### Disabled rules

The following rules are currently disabled. The rules and corresponding code which fails these rules can be inspected at a later stage and adjusted as needed:

https://github.com/cypress-io/cypress-realworld-app/blob/d23eba82a98d1bbf2cb72bbdf8b4b28dfba8666c/eslint.config.mjs#L22-L39
